### ISSUE TITLE
The validition accuracy is inconsistent with all epochs during the training process.

### ISSUE THREAD

**Issue by cainiao123s** [ultralytics/ultralytics#22711]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

In subsequent research, I found that the final results obtained from best.pt, whether it was map50 or map50-95, were inconsistent with all previous epochs. The model parameters during validation were slightly lower than those during training, so I suspect that some minor adjustments were made to the model during validation.

from ultralytics.models.yolo.detect import DetectionTrainer

if __name__ == '__main__':

    args = dict(model=r'yolo11n.pt', data=r'F:\j.yaml', epochs=3, batch=32, device='0',
                patience=0, )
    trainer = DetectionTrainer(overrides=args)
    train_generator = trainer._do_train()
D:\anaconda\envs\jacket_paltform\python.exe C:/Users/upc/AppData/Roaming/JetBrains/PyCharm2021.3/scratches/scratch_28.py
Ultralytics 8.3.99 🚀 Python-3.8.20 torch-2.0.1 CUDA:0 (NVIDIA GeForce RTX 4090, 24563MiB)
engine\trainer: task=detect, mode=train, model=yolo11n.pt, data=F:\j.yaml, epochs=3, time=None, patience=0, batch=32, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=train121, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=True, opset=None, workspace=None, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, copy_paste_mode=flip, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\train121
Overriding model.yaml nc=80 with nc=2

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      6640  ultralytics.nn.modules.block.C3k2            [32, 64, 1, False, 0.25]      
  3                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
  4                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  5                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  6                  -1  1     87040  ultralytics.nn.modules.block.C3k2            [128, 128, 1, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    346112  ultralytics.nn.modules.block.C3k2            [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1    249728  ultralytics.nn.modules.block.C2PSA           [256, 256, 1]                 
 11                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 12             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 13                  -1  1    111296  ultralytics.nn.modules.block.C3k2            [384, 128, 1, False]          
 14                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 15             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 16                  -1  1     32096  ultralytics.nn.modules.block.C3k2            [256, 64, 1, False]           
 17                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 18            [-1, 13]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 19                  -1  1     86720  ultralytics.nn.modules.block.C3k2            [192, 128, 1, False]          
 20                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 21            [-1, 10]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 22                  -1  1    378880  ultralytics.nn.modules.block.C3k2            [384, 256, 1, True]           
 23        [16, 19, 22]  1    431062  ultralytics.nn.modules.head.Detect           [2, [64, 128, 256]]           
YOLO11n summary: 181 layers, 2,590,230 parameters, 2,590,214 gradients, 6.4 GFLOPs

Transferred 448/499 items from pretrained weights
Freezing layer 'model.23.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks...
AMP: checks passed ✅
train: Scanning F:\导管架平台-损伤\训练用数据集\labels\train.cache... 802 images, 81 backgrounds, 0 corrupt: 100%|██████████| 883/883 [00:00<?, ?it/s]
val: Scanning F:\导管架平台-损伤\训练用数据集\labels\val.cache... 192 images, 29 backgrounds, 0 corrupt: 100%|██████████| 221/221 [00:00<?, ?it/s]
Plotting labels to runs\detect\train121\labels.jpg... 
  0%|          | 0/28 [00:00<?, ?it/s]optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to runs\detect\train121
Starting training for 3 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/3      4.52G      1.727      3.331      1.517        157        640: 100%|██████████| 28/28 [00:25<00:00,  1.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  5.05it/s]
                   all        221        807   0.006052     0.5067     0.2272     0.1087

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        2/3      4.53G      1.555       2.01      1.299        128        640: 100%|██████████| 28/28 [00:14<00:00,  1.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  6.03it/s]
                   all        221        807   0.006202     0.5141     0.2538     0.1446

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        3/3      4.53G       1.41      1.671      1.201        126        640: 100%|██████████| 28/28 [00:17<00:00,  1.60it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:00<00:00,  6.36it/s]
                   all        221        807     0.9007     0.2568     0.4796     0.3028

3 epochs completed in 0.017 hours.
Optimizer stripped from runs\detect\train121\weights\last.pt, 5.5MB
Optimizer stripped from runs\detect\train121\weights\best.pt, 5.5MB

Validating runs\detect\train121\weights\best.pt...
Ultralytics 8.3.99 🚀 Python-3.8.20 torch-2.0.1 CUDA:0 (NVIDIA GeForce RTX 4090, 24563MiB)
YOLO11n summary (fused): 100 layers, 2,582,542 parameters, 0 gradients, 6.3 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.88it/s]
                   all        221        807     0.9008      0.257     0.4794     0.3027
           Deformation        125        513     0.8648      0.212     0.4545     0.2567
                 Crack        107        294     0.9367      0.302     0.5044     0.3486
Speed: 0.5ms preprocess, 0.5ms inference, 0.0ms loss, 1.7ms postprocess per image
Results saved to runs\detect\train121

Process finished with exit code 0

### Additional

[results.csv](https://github.com/user-attachments/files/23595641/results.csv)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cainiao123s, thank you for opening this issue and for sharing detailed logs and `results.csv` 🚀  
This is an automated response to help you get support more quickly — an Ultralytics engineer will also review your report and assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for general usage guidance, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples and where many common questions are already answered ✅

Since your report concerns metrics and training/validation behavior, it appears to be a 🐛 **Bug Report**. To help us investigate, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) if possible, including:
- Exact training/validation commands or script (minimal version)
- Dataset structure description (number of images, classes) and any special settings in your `j.yaml`
- The steps needed to reproduce the inconsistency in metrics across epochs vs. the final `best.pt` evaluation

If this turns out to be more of a custom training ❓ **Question** than a bug, please also include:
- Dataset image examples (a few representative samples)
- Full training logs (or longer excerpts around the epochs of interest)
- Confirmation that you are following the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) 💡

Join the Ultralytics community where it suits you best:
- For real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧  
- For in-depth discussions: [Discourse](https://community.ultralytics.com/)  
- For community threads and sharing results: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Before we dive deeper, please upgrade to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/), and then check if the issue persists:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

When you reply, it will help a lot if you can confirm:
1. That you’ve upgraded to the latest `ultralytics` version and rerun the training/validation  
2. Whether the same behavior occurs when using the standard `yolo detect train ...` / `yolo detect val ...` commands instead of a custom script  
3. A minimal script or command that reproduces the discrepancy

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
The final checkpoint is saved as FP16. So there will be slight differences:

https://github.com/ultralytics/ultralytics/blob/29448b21a9494eb887570ce7d7f3cc1ad8c5b49b/ultralytics/utils/torch_utils.py#L709
